HW2/P2/parallel_vector.pyx

```diff
@@ -105,11 +139,42 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    # to parallelize data movement.  Be sure to avoid deadlock, as well as
    # double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(num_locks, locks)
+    with nogil:
+        for r in range(repeat):
+            for idx in prange(src.shape[0], num_threads=4):
+
+                # Same code as above, except now we wrap our indices around
+                si = src[idx] % num_locks
+                di = dest[idx] % num_locks
+
+                # Lock smallest first, and only lock once if theyre the same
+                same_lock = 0
+                if si < di:
+                    first_lock = si
+                    second_lock = di
+                elif di < si:
+                    first_lock = di
+                    second_lock = si
+                else:
+                    first_lock = si
+                    same_lock = 1
+
+                if counts[src[idx]] > 0:
+
+                   # Get our locks
+                    acquire(&locks[first_lock])
+                    if not same_lock:
+                        acquire(&locks[second_lock])
+
+                   # Increment/decrement accordingly
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
+
+                   # And now release the beast from its cage
+                    if not same_lock:
+                        release(&locks[second_lock])
+                        release(&locks[first_lock])
+                    else:
+                        release(&locks[first_lock])
+
+    free_N_locks(num_locks, locks)
```

Comment:
Nice! All looks good to me.  Not sure what would happen if you flipped the for loops around, so you had the for...prange wrapping the for...range loop, might be interesting to see if that was any faster.